shapeshift/web
# feat(cosmos send): add memo support

## Description

This adds support for memos in Cosmos (SDK) send modals

- display a memo field
- support a maximum memo length of 256 chars: there is no real spec/limitation in Cosmos SDK for this - it could be any go uint64 so technically up to 18,446,744,073,709,551,615 , but the auth module uses 256 as an example for `MaxMemoCharacters` https://github.com/cosmos/cosmos-sdk/blob/2627a96a496e3c47d97b5c34720a7359600b70ed/x/auth/spec/06_params.md
- shows the amount of remaining characters and doesn't let the user continue with the flow if the limit is exceeded
- on the next modal, the memo is displayed when confirming the transaction


## Notice




## Pull Request Type


## Issue (if applicable)

#651

## Risk

N/A

## Testing

- enable the Cosmos flags
- go to the ATOM (Cosmos) asset
- try to send some ATOM and add a memo

- memo should be accepted and it should be possible to continue the flow if it's less or equal to 256 chars
- memo should not be accepted and it should be impossible to continue the flow if 256 chars are exceeded - continue button is disabled, and in errored state withand the characters counter turns red
- any string should work as long as it is less or equal to 256 chars - tested using keplr and checking cosmos-sdk docs and code, there is no check except than this one 

## Screenshots (if applicable)

<img width="485" alt="image" src="https://user-images.githubusercontent.com/17035424/159542711-1997e813-6369-4d39-a41f-702a49c8163e.png">

<img width="471" alt="image" src="https://user-images.githubusercontent.com/17035424/159539161-dbfd0cf6-816f-4017-aaf5-a58bb581927b.png">
﻿
<img width="492" alt="image" src="https://user-images.githubusercontent.com/17035424/159539209-5bc11e02-72c5-46f2-bbe8-7e9d381572a0.png">

<img width="469" alt="image" src="https://user-images.githubusercontent.com/17035424/159539290-4308fa46-0e3b-4793-80f3-2245fa44394c.png">

<img width="462" alt="image" src="https://user-images.githubusercontent.com/17035424/159539309-fce49633-26f7-4b56-b67c-2ff16e0a1358.png">
